### ISSUE TITLE
[GatsbyJS] This component must be used inside a <RecoilRoot> component error in build

### ISSUE THREAD

**Issue by phlex89** [facebookexperimental/Recoil#486]
I'm trying to use Recoil Js in a GatsbyJs application.

I configured `<RecoilRoot>` in `wrapRootElement`  (also tried `import { RecoilRoot } from 'recoil/dist/recoil.production' `  as suggested in [this issue](https://github.com/facebookexperimental/Recoil/issues/301)) but I receive this error during `gatsby build` script.

Am  I missing something? How to use this lib in a GatsbyJs app?

**Comment by mrlubos** [comment]
Hi @phlex89, do you have an example to reproduce this issue please?

**Comment by luis-aguilar-raweng** [comment]
Hi guys, any news on this issue I believe I am experiencing the same trying to use RecoilRoot in the wrapRootElement. I think I can provide an example of my problem if that's ok, I don't want to hijack this issue.

**Comment by luis-aguilar-raweng** [comment]
Hi again, I was actually able to solve by just adding the RecoilRoot to the gatsby-ssr file, in wrapRootElement. Hope it helps!

**Comment by risakanagawa** [comment]
@luis-aguilar-raweng 
Hi I'm having this issue as well. Could I have a detail example code?

**Comment by luis-aguilar-raweng** [comment]
@risakanagawa For sure, hope this helps. 

First, know that the gatsby-browser file "...lets you...wrap your site in additional components..." - https://www.gatsbyjs.com/docs/browser-apis. You need to wrap your root element to allow passing the recoil state, for example:

```
exports.wrapRootElement = ({ element }) => {
  
  return (
      <RecoilRoot>
         { element }
      </RecoilRoot>
  )
}
```

Second, if you have the gatsby-ssr file then you pretty much need to do the same since "...There is an equivalent hook in Gatsby’s Browser API. It is recommended to use both APIs together." - https://www.gatsbyjs.com/docs/ssr-apis/

**Comment by risakanagawa** [comment]
Thank you for the reply @luis-aguilar-raweng !
I realized I've mixed with common JS and ES6 which probably made this happen.
I just stick with ES6 syntax `import` and `export` which solved my error but definitely your comment inspired me! 
Really appreciated  😄

**Comment by WiredEntrepreneur** [comment]
Thanks for the info. The following links referenced helped me resolve this at my end.

https://www.gatsbyjs.com/docs/ssr-apis/#wrapRootElement

https://www.gatsbyjs.com/docs/browser-apis/#wrapRootElement

Added this in both the files:

const React = require('react');
const { RecoilRoot } = require('recoil');
exports.wrapRootElement = ({ element }) => {  
    return (
        <RecoilRoot>
           { element }
        </RecoilRoot>
    )
  }

**Comment by BayanBennett** [comment]
Just throwing this on here just in case anyone is stuck and they happen to have the same problem that I did.

### ❌Doesn't work
```js
module.exports = {
  wrapRootElement: ({ element }) =>
    createElement(RecoilRoot, null, element);
}

```

### ✔ Works
```js
module.exports.wrapRootElement = ({ element }) =>
    createElement(RecoilRoot, null, element);
```

**Comment by lukehillonline** [comment]
Hey All,

Sorry to bring this up again after it has been closed but I am having an issue getting Recoil up and running with Gatsby and I am not sure what I am doing wrong.

I have created a sandbox that started off as the Gatsby Starter Kit and then I added recoil. I've put in RecoilRoot through the gatsby-browser and gatsby-ssr files and then I am using my small bit of state in pages/index.js

Here is a link:
https://codesandbox.io/s/cool-cache-sdfzc?file=/gatsby-browser.js

I think I am doing everything to the documentation so I was hoping someone could take a look and see where I am going wrong.

Thanks in advance.
